contracts/core/OTCWrapper.sol

```diff
@@ -0,0 +1,690 @@
+        require(whitelist.isWhitelistedCollateral(_collateralAsset), "OTCWrapper: collateral is not whitelisted");
+
+        uint256 orderFee = (order.notional.mul(fee[order.underlying])).div(10000);
+
+        uint256 totalInitialMargin = _collateralAmount;
+        if (_collateralAsset == USDC) {
+            totalInitialMargin = _collateralAmount.add(_premium).sub(orderFee);
+        }
+
+        require(
+            marginRequirements.checkMintCollateral(
+                msg.sender,
+                order.notional,
+                order.underlying,
+                order.isPut,
+                totalInitialMargin,
+                _collateralAsset
+            ),
+            "OTCWrapper: insufficient collateral"
+        );
+
+        // user inflow
+        _depositWithPermit(
+            _userSignature.acct,
+            USDC,
+            _premium,
+            _userSignature.deadline,
+            _userSignature.v,
+            _userSignature.r,
+            _userSignature.s
+        );
+
+        // market maker inflow
+        _depositWithPermit(
+            _mmSignature.acct,
+            _collateralAsset,
+            _collateralAmount,
+            _mmSignature.deadline,
+            _mmSignature.v,
+            _mmSignature.r,
+            _mmSignature.s
+        );
+
+        // transfer fee to beneficiary address
+        IERC20(USDC).safeTransfer(beneficiary, orderFee);
+
+        // transfer premium to market maker only if collateral is not USDC
+        if (_collateralAsset != USDC) {
+            IERC20(USDC).safeTransfer(msg.sender, _premium.sub(orderFee));
+        }
+
+        // open vault
+        uint256 vaultID = (controller.getAccountVaultCounter(address(this))).add(1);
+
+        UtilsWrapperInterface.ActionArgs[] memory actions = new UtilsWrapperInterface.ActionArgs[](3);
+
+        actions[0] = UtilsWrapperInterface.ActionArgs(
+            UtilsWrapperInterface.ActionType.OpenVault,
+            address(this), // owner
+            address(this), // receiver
+            address(0), // asset, otoken
+            vaultID, // vaultId
+            0, // amount
+            0, // index
+            abi.encode(1) // vault type
+        );
+
+        // Approve margin pool to deposit collateral
+        IERC20(_collateralAsset).safeApproveNonCompliant(addressbook.getMarginPool(), totalInitialMargin);
+
+        // deposit collateral
+        actions[1] = UtilsWrapperInterface.ActionArgs(
+            UtilsWrapperInterface.ActionType.DepositCollateral,
+            address(this), // owner
+            address(this), // address to transfer from
+            _collateralAsset, // deposited asset
+            vaultID, // vaultId
+            totalInitialMargin, // amount
+            0, //index
+            "" //data
+        );
+
+        // create otoken, get an address and calculate mint amount
+        address otoken = IOtokenFactoryWrapperInterface(addressbook.getOtokenFactory()).createOtoken(
```

Comment:
you should do something similar to [getOrDeployOtoken()](https://github.com/ribbon-finance/ribbon-v2/blob/e9270281c7aa7433851ecee7f326c37bce28aec1/contracts/libraries/VaultLifecycle.sol#L576) to make sure it does not already exist, or else it will revert

Reply:
https://github.com/ribbon-finance/GammaProtocol-OTC/pull/2/commits/3491cd25dd2fe74965cd8053e2de617a5e3bb5ad